planner/optimize.go

```diff
@@ -75,6 +102,111 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 	return plannercore.DoOptimize(ctx, builder.GetOptFlag(), logic)
 }
 
+func addHint(ctx sessionctx.Context, stmtNode ast.StmtNode) (ast.StmtNode, *bindinfo.HintsSet) {
+	// When the domain is initializing, the bind will be nil.
+	if ctx.Value(bindinfo.SessionBindInfoKeyType) == nil {
+		return stmtNode, nil
+	}
+	switch x := stmtNode.(type) {
+	case *ast.ExplainStmt:
+		var oriHintSet *bindinfo.HintsSet
+		switch x.Stmt.(type) {
+		case *ast.SelectStmt:
+			normalizeExplainSQL := parser.Normalize(x.Text())
+			idx := strings.Index(normalizeExplainSQL, "select")
+			normalizeSQL := normalizeExplainSQL[idx:]
+			hash := parser.DigestHash(normalizeSQL)
+			x.Stmt, oriHintSet = addHintForSelect(hash, normalizeSQL, ctx, x.Stmt)
+		}
+		return x, oriHintSet
+	case *ast.SelectStmt:
+		normalizeSQL, hash := parser.NormalizeDigest(x.Text())
+		return addHintForSelect(hash, normalizeSQL, ctx, x)
+	default:
+		return stmtNode, nil
+	}
+}
+
+func addHintForSelect(hash, normdOrigSQL string, ctx sessionctx.Context, stmt ast.StmtNode) (ast.StmtNode, *bindinfo.HintsSet) {
+	sessionHandle := ctx.Value(bindinfo.SessionBindInfoKeyType).(*bindinfo.SessionHandle)
+	bindRecord := sessionHandle.GetBindRecord(normdOrigSQL, ctx.GetSessionVars().CurrentDB)
+	if bindRecord != nil {
+		if bindRecord.Status == bindinfo.Invalid {
+			return stmt, nil
+		}
+		if bindRecord.Status == bindinfo.Using {
+			metrics.BindUsageCounter.WithLabelValues(metrics.ScopeSession).Inc()
+			oriHint := bindinfo.CollectHint(stmt)
+			return bindinfo.BindHint(stmt, bindRecord.HintsSet), oriHint
+		}
+	}
+	globalHandle := domain.GetDomain(ctx).BindHandle()
+	bindRecord = globalHandle.GetBindRecord(hash, normdOrigSQL, ctx.GetSessionVars().CurrentDB)
+	if bindRecord == nil {
+		bindRecord = globalHandle.GetBindRecord(hash, normdOrigSQL, "")
+	}
+	if bindRecord != nil {
+		metrics.BindUsageCounter.WithLabelValues(metrics.ScopeGlobal).Inc()
+		oriHint := bindinfo.CollectHint(stmt)
+		return bindinfo.BindHint(stmt, bindRecord.HintsSet), oriHint
+	}
+	return stmt, nil
+}
+
+func handleInvalidBindRecord(ctx context.Context, sctx sessionctx.Context, stmtNode ast.StmtNode) {
+	var normdOrigSQL, hash string
+	switch x := stmtNode.(type) {
```

Comment:
There are already two places of this logic to gain `normOrigSQL` and `hash`. Can we extract a function for it?